kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -499,348 +532,174 @@ protected void handleInactive(KafkaHeaderAndRequest kafkaHeaderAndRequest,
-                                                        ignored -> Collections.synchronizedList(new ArrayList<>())
-                                                ).add(topicName);
-                                            }
-                                            completeOne.run();
-                                        });
-                            } else {
-                                completeOne.run();
-                            }
-                        });
             }
-        }).exceptionally(error -> {
-            topicMapFuture.completeExceptionally(error);
-            return null;
         });
-        return topicMapFuture;
+        return topicAndMetadataList;
+    }
+
+    private CompletableFuture<List<String>> authorizeNamespacesAsync(final Collection<String> namespaces,
+                                                                     final AclOperation aclOperation) {
+        final Map<String, CompletableFuture<Boolean>> futureMap = namespaces.stream().collect(
+                Collectors.toMap(
+                        namespace -> namespace,
+                        namespace -> authorize(aclOperation, Resource.of(ResourceType.NAMESPACE, namespace))
+                ));
+        return CoreUtils.waitForAll(futureMap.values()).thenApply(__ ->
+            futureMap.entrySet().stream().filter(e -> {
+                if (!e.getValue().join()) {
+                    log.warn("Failed to authorize {} for ACL operation {}", e.getKey(), aclOperation);
+                    return false;
+                }
+                return true;
+            }).map(Map.Entry::getKey).collect(Collectors.toList())
+        );
+    }
+
+    private CompletableFuture<Stream<String>> listAllTopicsFromNamespacesAsync(final List<String> namespaces) {
+        return CoreUtils.waitForAll(namespaces.stream()
+                .map(namespace -> pulsarService.getNamespaceService()
+                        .getListOfPersistentTopics(NamespaceName.get(namespace))
+                ).collect(Collectors.toList()),
+                topics -> topics.stream().flatMap(List::stream));
+    }
+
+    private CompletableFuture<ListPair<String>> authorizeTopicsAsync(final Collection<String> topics,
+                                                                     final AclOperation aclOperation) {
+        final Map<String, CompletableFuture<Boolean>> futureMap = topics.stream().collect(
+                Collectors.toMap(
+                        topic -> topic,
+                        topic -> authorize(aclOperation, Resource.of(ResourceType.TOPIC, topic))
+                ));
+        return CoreUtils.waitForAll(futureMap.values()).thenApply(__ ->
+                ListPair.of(futureMap.entrySet()
+                        .stream()
+                        .collect(Collectors.groupingBy(e -> e.getValue().join()))
+                ).map(Map.Entry::getKey));
+    }
```

Comment:
We should handle exceptions that occur when checking permissions. I see that `io.streamnative.pulsar.handlers.kop.KafkaRequestHandler#authorize()` may throw exception.

Reply:
`authorize` doesn't throw an exception, instead, the future will be completed exceptionally. In this case, the future returned by `CoreUtils.waitForAll` will be completed exceptionally. The exception will eventually be processed in `getTopicsAsync`'s `whenComplete`.

It should be noted that there are two failed cases:
- `authorize` returns false, which means the topic doesn't have the expected permission. In this case, an `AUTHORIZATION_FAILURE` will be set.
- `authorize` failed with exception. In this case, we treat it as an unrecoverable failure and prevent the followed operations.

See https://github.com/streamnative/kop/issues/1105. This PR changes the current behavior for the 2nd case.

Reply:
One of my concerns is that if handle permission validation with an unrecoverable error, instead of having no permissions, the client won't be able to receive the exception from the server, even though it doesn't care about the exception, but since the client just got an empty TopicMetadata , the client will continue to retry the `MetadataRequest` , and a large number of `MetadataRequest` will consume more CPU of the broker. In addition, the client may not find that there is a problem with its own program. 

I looked at the implementation of `handleMetadataRequest` in kafka. If an broker level fault is encountered, the broker will return the error to the client.